call.go

```diff
@@ -0,0 +1,98 @@
+package rpc
+
+import (
+	"context"
+	"sync"
+
+	inet "github.com/libp2p/go-libp2p-net"
+	peer "github.com/libp2p/go-libp2p-peer"
+)
+
+// Call represents an active RPC. Calls are used to indicate completion
+// of RPC requests and are returned within the provided channel in
+// the Go() functions.
+type Call struct {
+	ctx    context.Context
+	cancel func()
+
+	finishedMu sync.RWMutex
+	finished   bool
+
+	Dest  peer.ID
+	SvcID ServiceID   // The name of the service and method to call.
+	Args  interface{} // The argument to the function (*struct).
+	Reply interface{} // The reply from the function (*struct).
+	Done  chan *Call  // Strobes when call is complete.
+
+	errorMu sync.Mutex
+	Error   error // After completion, the error status.
+
+}
+
+func newCall(ctx context.Context, dest peer.ID, svcName, svcMethod string, args interface{}, reply interface{}, done chan *Call) *Call {
+	ctx2, cancel := context.WithCancel(ctx)
+	return &Call{
+		ctx:    ctx2,
+		cancel: cancel,
+		Dest:   dest,
+		SvcID:  ServiceID{svcName, svcMethod},
+		Args:   args,
+		Reply:  reply,
+		Error:  nil,
+		Done:   done,
+	}
+}
+
+// done places the completed call in the done channel.
+func (call *Call) done() {
+	call.finishedMu.Lock()
+	call.finished = true
+	call.finishedMu.Unlock()
+
+	select {
+	case call.Done <- call:
+		// ok
+	default:
+		logger.Debugf("discarding %s.%s call reply",
+			call.SvcID.Name, call.SvcID.Method)
+	}
+	call.cancel()
+}
+
+func (call *Call) doneWithError(err error) {
+	if err != nil {
+		logger.Error(err)
+		call.setError(err)
+	}
+	call.done()
+}
+
+func (call *Call) isFinished() bool {
+	call.finishedMu.RLock()
+	defer call.finishedMu.RUnlock()
+	return call.finished
```

Comment:
Yes 👍, my question wasn't phrased too well and I cleared up my confusion

Reply:
@ZenGround0 yes, base types are copies. If it was a map, it would essentially be a pointer, so the lock would be useless if what you were trying to lock were the things the pointer is referencing (at not the pointer itself). That's how I understand it.